[Playlists] Manual Model + API updates

bjtitus opened the pull request:
Fixes [PCIOS-117](https://linear.app/a8c/issue/PCIOS-117/manual-playlist-db-model-sync)

* Removes the old `rawPlaylistType` type and DB column
* Adds the `manual` property to our local model (the column already existed)
* Update the protobuf file with the latest models including the added fields for manual playlist sync: `episodes` and `episodeOrder` on `Api_SyncUserPlaylist` and `Api_PlaylistSyncResponse` as well as the `Api_SyncPlaylistEpisode` which is a stub with barebones episode info.
* Updated `SyncTask+LocalChanges.swift` and `SyncTask+ServerChanges.swift` with conversion to/from these protobuf models
* Unifies the `playlistEpisodes` and `playlistEpisodeCount` methods. The queries can diverge deeper down and remove the extra methods

## To test

### Create
* Enable the `playlistsRebranding` Feature Flag
* Add some episodes to the playlist both from your library and from Discover.
* Create a Manual playlist
* ✅ Ensure that all episodes from Up Next are added to the playlist

### Sync
* Sign in with the same account on two devices
* Create a playlist (this currently uses the contents of your Up Next queue)
* Refresh the profile
* Ensure the playlist is synced over
* Delete the playlist and sync
* ✅ Ensure the playlist is removed from both devices

### Missing episodes
* Add an episode to the Up Next queue from Discover
* Create a Manual playlist (all Up Next episodes will be added)
* Delete that episode from the Up Next queue so it isn't synced as part of another process
* Sync the account
* ✅ Ensure that the playlist shows up on the other device and the episode removed from Up Next is working

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

bjtitus committed:
Update protobuf api file with playlist updates

bjtitus committed:
Update query methods + data models for manual playlist

bjtitus committed:
Add Up Next queue as initial set of episodes for testing

bjtitus committed:
Switch from playlistType to manual property

This matches Android and better matches the protobuf model

bjtitus committed:
Remove separate manual/smart playlist methods

These are mostly the same and can be branched within the Query

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pocketcasts commented:
<!-- REUSE_ID: c9f9bbbb-793b-48b6-a044-2a4c2ea1f056 -->

Version `7.98` has now entered code-freeze, so the milestone of this PR has been updated to `7.99`.

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Public/PlaylistQueryBuilder.swift:
```diff
@@ -35,7 +35,8 @@ public class PlaylistQueryBuilder {
         let select = select(clause: clause)
         let addedUuid = add(episodeUuidToAdd: episodeUuidToAdd)
         let rules = add(smartRulesFor: playlist)
```
@bjtitus are we sure that this builder is ok to be used for manual playlists? For example you didn't filter the smart rules method, which is not necessary for manual playlist. Also there's an extra `AND` when filtering out the rules when fetching a manual playlist like (which generates an error):
```
SELECT episode.* FROM SJEpisode episode LEFT JOIN SJPodcast podcast ON episode.podcast_id = podcast.id WHERE episode.archived = 0 AND ( ) AND episode.uuid IN ('e92dfec3-0b3f-412c-b63d-236fe8d4318c') ORDER BY episode.publishedDate DESC, episode.addedDate DESC  LIMIT 1000
```
Compared to the Android query for manual, down need to do something else or is this enough?

danielebogo reviewed (changes requested):
@bjtitus the changes seem good but I found few things:

• I left a comment below as Im not sure the query used is correct. A part from my comment of having to remove the smart rules, if I manually remove them and create a new manual playlist I always have an episode added.

• Do you have other tests we can do to test all these changes?

bjtitus committed:
Add unit tests

bjtitus committed:
Improve SQL query building

I think this could still use improvement but it's a start to make the builder more flexible.

bjtitus committed:
Fix server syncing

bjtitus committed:
Update Unit Test

bjtitus committed:
Remove episodes column from playlist

bjtitus committed:
Update queries for Playlist episodes

bjtitus committed:
Fix playlist episode queries + unit tests

bjtitus committed:
Update episode saving to filters

bjtitus committed:
Filter for up next playlist_id in UpNextDataManager

bjtitus committed:
Update episodes query to avoid duplication

bjtitus committed:
Remove episodes alongside filters

bjtitus committed:
Update episode + podcast table on new playlist episodes

bjtitus committed:
Merge remote-tracking branch 'origin/trunk' into bjtitus/playlists/manual-api-updates

# Conflicts:
#	Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeFilterDataManager.swift
#	Modules/DataModel/Sources/PocketCastsDataModel/Public/DataManager.swift
#	Modules/Server/Sources/PocketCastsServer/Private/API Tasks/RetrievePlaylistsTask.swift
#	Modules/Server/Sources/PocketCastsServer/Public/Sync/SyncTask+LocalChanges.swift
#	Modules/Server/Sources/PocketCastsServer/Public/Sync/SyncTask+ServerChanges.swift
#	podcasts/New Creation/New Playlist/NewPlaylistViewController.swift

bjtitus committed:
Update Ghost Episode find to ignore playlist episodes

These episodes need to stay even if they go missing

bjtitus committed:
Add check for playlist when cleaning up episodes

Episodes which are in a playlist need to be left. We will need to update the UI as a follow up

bjtitus committed:
Update unit tests

bjtitus committed:
Fix trailing newline

bjtitus committed:
Add episodes on server sync before updating

This fixes synchronous unit tests but also ensures we have an entry in the case the episode fetch fails

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/PlaylistDataManager.swift:
```diff
@@ -233,35 +232,92 @@ class PlaylistDataManager {
+        filter.shorterThan = rs.int(forColumn: "shorterThan")
+        filter.manual = rs.bool(forColumn: "manual")
+
+        return filter
```
The var name should be `playlist`

danielebogo commented on Modules/Server/Sources/PocketCastsServer/Public/Sync/SyncTask+LocalChanges.swift:
```diff
@@ -118,37 +118,75 @@ extension SyncTask {
+            }
+            filterRecord.episodeOrder = episodes.map { $0.uuid }
+        }
+        return filterRecord
```
We should use `playlistRecord`

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeDataManager.swift:
```diff
@@ -45,6 +45,12 @@ class EpisodeDataManager {
 
+    enum Constants {
+        enum Limits {
+            static let maxFilterItems = FeatureFlag.playlistsRebranding.enabled ? 1000 : 500
```
Can this be renamed as `maxPlaylistItems`?

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeDataManager.swift:
```diff
@@ -1007,6 +1040,28 @@ class EpisodeDataManager {
         }
     }
 
+    func episodes(for filter: EpisodeFilter, limit: Int = Constants.Limits.maxFilterItems, dbQueue: PCDBQueue) -> [Episode] {
```
```suggestion
    func episodes(for playlist: EpisodeFilter, limit: Int = Constants.Limits.maxFilterItems, dbQueue: PCDBQueue) -> [Episode] {
```

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeDataManager.swift:
```diff
@@ -1007,6 +1040,28 @@ class EpisodeDataManager {
 
+    func episodes(for filter: EpisodeFilter, limit: Int = Constants.Limits.maxFilterItems, dbQueue: PCDBQueue) -> [Episode] {
+
+        let query = PlaylistQueryBuilder.queryFor(filter: filter, episodeUuidToAdd: nil, limit: limit)
```
This is using the legacy method of the `PlaylistQueryBuilder`, to fetch only smart playlists episodes

bjtitus replied:
This method isn't being used any more. I think I may have accidentally added it back as part of the merge with the playlist rename.

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/EpisodeDataManager.swift:
```diff
@@ -1007,6 +1040,28 @@ class EpisodeDataManager {
+                    }
+                }
+            } catch {
+                FileLog.shared.addMessage("EpisodeFilterDataManager.episodesForFilter error: \(error)")
```
This should be `EpisodeDataManager` and the method should reflect the name with no filter

danielebogo reviewed (changes requested):
Thanks @bjtitus for all these changes!

I left few comments but overall it's an mazing addition.

Few things:
- The manual creation worked. I see that the manual playlist though used the up next queue episodes. Is it for testing only?
- After adding that manual playlist, logging into a new devices took few minutes and tentatives to have the manual playlist back. Also I didn't receive the smart playlists. Is it expected?

bjtitus committed:
Filter → Playlist renames

bjtitus committed:
Remove unused `episodes(for filter)`

bjtitus commented:
@danielebogo 

> I see that the manual playlist though used the up next queue episodes. Is it for testing only?

Yes, [this is just for testing](https://github.com/Automattic/pocket-casts-ios/pull/3491/files#diff-9d36bac2972b6f7f38cd1bf5d4a17e834f67cc7adf0e598f6bf04587ecb7f43dR146-R150) at the moment since we don't have the UI for adding episodes to a playlist. Do you think I should do anything else to call this out?

> I didn't receive the smart playlists. Is it expected?

It _should_ sync Smart Playlists as well although I definitely tested Manual playlists more. Could you send a recording?

https://github.com/user-attachments/assets/d6fab07d-40e7-445f-93ce-cc90c4a23cfc

danielebogo commented:
@bjtitus I did this test today:

### Start with Trunk
1. Checkout trunk
2. New install on my real device 
3. Login with my user
4. I see my Smart Playlists pulled
5. I install trunk as new install on my sim SE
6. Login with the same user
7. I get all the Smart Playlists
8. On my Real device I create a new smart Playlist
9. I reload the user on the sim SE and I get the new Smart

### Switch to this branch
1. Install this branch over trunk on my real phone
2. I still see the previous Smart Playlist
3. I create a new Manual Playlist
4. I delete the app on the sim SE
5. Install your branch, sign in with the same user
6. The playlists section is empty. It didn't sync any Playlists
7. Trying to reload the the user
8. Still empty Playlists

bjtitus committed:
Fix playlist task fetching on full sync

bjtitus committed:
Add Unit Tests for playlist list methods

danielebogo reviewed (approved):
Perfect! it's working properly! :shipit:

bjtitus committed:
Merge branch 'trunk' into bjtitus/playlists/manual-api-updates

# Conflicts:
#	Modules/Server/Sources/PocketCastsServer/Private/Protobuffer/api.pb.swift
